File: 08.py
#!/usr/bin/env python3
from sys import stdin
from itertools import permutations as P

chars  = "abcdefg"
nums   = ["abcefg","cf","acdeg","acdfg","bcdf","abdfg","abdefg","acf","abcdefg","abcdfg"]
nums   = list(map(set,nums))
digits = lambda x: [i for i,n in enumerate(nums) if n==set(x)]
t      = lambda x,ii: "".join([chars[ii[chars.find(c)]] for c in x])
rs     = list(map(lambda l: l.replace('| ','').split(), stdin.readlines()))
conv   = lambda p, xs: sum((digits(t(x,p)) for x in xs),[])
find   = lambda xs: next(p for p,i in ((p,len(conv(p,xs))) for p in P([0,1,2,3,4,5,6])) if i==10)
ps     = [conv(p, r[10:]) for r,p in ((r, find(r[0:10])) for r in rs)]
p1     = len(list(filter(lambda x: x in [1,4,7,8], sum(ps,[]))))
p2     = sum(map(lambda x: int("".join(map(str,x))), ps))

print(f"Part 1: { p1 }, Part 2: { p2 }")

File: 08.scala
#!/usr/bin/env scala
import scala.io.Source.stdin

@main def aoc(): Unit =
  val ch = "abcdefg".toVector
  val ns = List("abcefg","cf","acdeg","acdfg","bcdf","abdfg","abdefg","acf","abcdefg","abcdfg").map(_.toSet)
  val xs = stdin.getLines().map(_.replace("| ","").split(" ").toList).toList
  val tr = (xs: String, p:IndexedSeq[Int]) => xs.map(x=>ch(p(ch.indexOf(x)))).toSet
  val pr = (xs:List[String], p:IndexedSeq[Int]) => xs.map(tr(_,p)).forall(ns.contains)
  val f  = (xs:List[String]) => (0 to 6).permutations.find(pr(xs,_)).get
  val ps = xs.map(x=>(x,f(x.dropRight(4)))).map((x,p)=>x.takeRight(4).map(i=>ns.indexOf(tr(i,p))))
  val p1 = ps.map(_.count(List(1,4,7,8).contains)).sum
  
  println(s"Part 1: ${ p1 }, Part 2: ${ ps.map(_.mkString.toInt).sum }")
